index.rst

```diff
@@ -51,11 +51,182 @@
+The client code would then use a local Datastore implementation to create the file from the relevant formatter and then upload it.
+The downside of this approach is that it is not immediately clear how to handle composite disassembly (where the file is split into multiple components and each is stored separately in the file store) since that is a datastore configuration.
+
+A ``Butler.get()`` would most logically be handled as a call to ``Butler.getURIs()`` (probably via the datastore ``getStoredItemsInfo`` method) to obtain the signed URL, download the file, and then use a local Datastore implementation to read that file.
+This could conceivably leverage the purported "caching datastore" concept.
+
+Even if all this is made to work, users still expect to be able to use many of the registry methods for querying the collections, datasets and dimension records.
+This suggests that it might be better to implement the client/server registry first and build on that, and then subsequently add explicit ``Butler`` overrides if performance is an issue (for example if ``Butler`` calls registry methods in a loop there may be significant run time overheads).
+
+Manager Classes
+---------------
+
+The final option is to implement each of the manager classes as client/server implementations.
+The registry is already designed to support pluggable managers and they are already designed to isolate database access.
+This seems like the cleanest way forward but the interfaces are very low level and this makes it significantly harder for non-Butler clients to do anything useful with the data being returned.
+It also raises the possibility of the interfaces being very slow when called in loops and may require significant caching of results and also the addition of new methods that move loops into the server.
+
+
+Access Controls
+===============
+
+Part of the benefit of using a client/server approach is that the server can control access to collections and datastores without having to use fine-grained database permissions on specific tables or add ACLs to the object store.
+
+This does though mean that there must be code in the server that can take the user name and determine which information can be used.
+This does not simply mean checking that the collection name includes the user name since the checks must also be able to look at collections that have group access controls (one person may wish to give access of their processed data to another user but no-one else).
+
+Conclusion
+==========
+
+After analyzing the options given above, the baseline plan is to:
+
+#. Support a pluggable registry in butler configuration and define a lightweight ABC interface.
+#. Write a cutdown client Registry that implements the more common methods.
+#. Implement a datastore subclass that can update datastore registry records via a server.
+#. Support butler put and get using these classes.
+#. Add URL signing to datastore.
```

Comment:
After my last big comment, I do think it makes sense to focus on implementing read-only versions of everything first, and to make that explicit here (even if we continue to talk about a design that would extend to read-write).